### PR TITLE
fix(qwen3-omni): auto-fallback to triton MoE runner on SM 80 (A100)

### PR DESCRIPTION
**Pull request by YuTaoV5**
Qwen3-Omni speech pipeline fails to start on NVIDIA A100 (SM 80) during CUDA graph capture of the talker_ar stage:

  File ".../flashinfer/fused_moe/core.py", line 331, in get_cutlass_fused_moe_module
    raise ValueError(f"Invalid backend: {backend}")
  ValueError: Invalid backend: 80

Root cause: sglang defaults the Qwen3-Omni MoE runner to flashinfer_cutlass, but flashinfer's cutlass fused_moe only supports SM 89, 90, 100, 103, 120. SM 80 (A100) is not in the list, and sglang's auto-resolution does not detect this gap.

Fix:
- Add sglang_omni/models/qwen3_omni/compat.py as a single point of truth for Qwen3-Omni hardware workarounds. Currently it exports get_qwen3_omni_compat_overrides() which returns {"moe_runner_backend": "triton"} only on SM 80 hosts, and an empty dict on every other GPU.
- Touch the two AR stage factories in stages.py (thinker and talker) to merge the compat overrides into their server_args_overrides dict before constructing ServerArgs.
- An escape hatch SGLANG_OMNI_FORCE_SM80_MOE_TRITON=1 is available for bisecting perf regressions or reproducing the flashinfer cutlass failure on non-A100 hardware.

Behavior contract:
- A100 (SM 80): moe_runner_backend="triton" (fix; was: ValueError)
- H100 / H20 (SM 90): unchanged (still flashinfer_cutlass)
- B100 / B200 (SM 100): unchanged
- RTX 5090 / Blackwell consumer (SM 120): unchanged
- Other Qwen3-Omni topologies (text-only, colocated, FP8): unaffected
- Other models (Higgs TTS, Fish Audio, Voxtral TTS, Qwen3-TTS, MOSS-TTS, Ming-Omni, LLaDA2-Uni, Whisper ASR, Qwen3-ASR): unaffected (only Qwen3-Omni imports compat)

Verification:
- Real A100 (8.0): overrides={"moe_runner_backend": "triton"} ✓
- Simulated H100 (9.0): overrides={} ✓ (matches upstream)
- End-to-end: 8-stage speech pipeline boots, 1.9s warm latency, 1.9s for 5 concurrent requests
- Unit tests: tests/unit_test/qwen3_omni/ (257 tests) collect OK, no regressions in the 33-test smoke subset

Refs: flashinfer cutlass fused_moe arch list at
  flashinfer/fused_moe/core.py (gen_*_module functions, supported archs: sm_89, sm_90, sm_100, sm_103, sm_120)

## Checklist

- [ ] Format your code according with pre-commit.
- [ ] Add unit tests.
- [ ] Update documentation / docstrings / example tutorials as needed.
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed.
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.

## CI

CI runs on self-hosted GPU runners and requires a maintainer to add the
`run-ci` label. Once labeled, every subsequent push re-triggers CI as
long as the label remains. Draft PRs are skipped even if labeled.

### CODE REVIEW

File: sglang_omni/models/qwen3_omni/compat.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    and the result is stable for the lifetime of the process.
+    """
+    try:
+        import torch
+    except ImportError:
+        return None
+    if not torch.cuda.is_available():
+        return None
+    try:
+        return torch.cuda.get_device_capability(0)
```
Comment (Hayden727) [inline sglang_omni/models/qwen3_omni/compat.py]: [P1] This should be keyed off the stage's actual `gpu_id`, not always CUDA device 0. The Qwen3-Omni factories pass `gpu_id` through to the scheduler, and the serve/config path supports placing thinker/talker on non-zero GPUs. On a mixed or non-isolated host, an A100 stage on `cuda:1` would miss the fallback when `cuda:0` is H100, while an H100/B100 stage could be forced to triton when `cuda:0` is A100. Please thread the factory `gpu_id` into the compat check and cache per device, with a small test that mocks different capabilities for device 0/1.

File: sglang_omni/models/qwen3_omni/stages.py [hunk 2]
```diff
@@ -942,10 +943,13 @@ def create_sglang_thinker_executor_from_config(
...
 ):
     """Returns OmniScheduler for thinker."""
-
     overrides: dict[str, Any] = {"disable_cuda_graph": False}
     if server_args_overrides:
         overrides.update(server_args_overrides)
+    # Hardware workarounds (e.g. SM 80 -> triton MoE). Applied last so
+    # they cannot be silently overridden by upstream kwargs, but still
+    # before the caller's explicit server_args_overrides is honored.
+    overrides.update(_qwen3_omni_compat.get_qwen3_omni_compat_overrides())
```
Comment (Hayden727) [inline sglang_omni/models/qwen3_omni/stages.py]: [P2] This merge happens after `server_args_overrides`, so the SM80 compat dict silently overwrites an explicit caller setting for `moe_runner_backend`. That breaks the existing explicit-override contract we already enforce in the model-worker backend-policy tests, and it also contradicts the comment above saying caller overrides are honored. Please only fill the compat value when the backend is absent/`auto`, or fail fast with a clear error if a specific explicit backend is unsupported on SM80. Same issue applies to the talker merge below.

**Review by Hayden727 — CHANGES_REQUESTED** [review]
Thanks for the A100 fix. I found two compatibility-contract issues that should be addressed before merge.

**Comment by YuTaoV5** [comment]
# PR #711 Update Notes After Hayden727 Review

> **PR**: https://github.com/sgl-project/sglang-omni/pull/711
> **Current commit**: `5ab798f`
> **Latest total diff**: +123 / -1 across 2 files

---

## 1. Issues Raised By Hayden727

### P1: `compat` always checked `cuda:0`, which is wrong on mixed-GPU placement

Original implementation:
```python
def _primary_device_capability() -> tuple[int, int] | None:
    return torch.cuda.get_device_capability(0)   # always device 0
```

Impact:
- With `cuda:0=H100 + cuda:1=A100`, a thinker on `cuda:0` (H100) could receive an unnecessary triton override and slow down by 5-15%, while a talker on `cuda:1` (A100) could miss the required triton override and crash.
- Qwen3-Omni placement supports non-zero GPUs, but the compat layer assumed single-device `cuda:0` placement.

### P2: caller-provided backend overrides were silently overwritten

Original implementation:
```python
overrides.update(_qwen3_omni_compat.get_qwen3_omni_compat_overrides())
```

Because `dict.update()` overwrites earlier values, this meant:
- A caller explicitly setting `moe_runner_backend="cutlass"` could be silently changed to `triton`, violating the explicit-override contract enforced by the model-worker backend-policy tests.
- The `stages.py` comment saying caller-provided `server_args_overrides` were honored was not true.

---

## 2. Fix

### File 1: `sglang_omni/models/qwen3_omni/compat.py` (new, 83 lines)

Changes:
- Replaced `_primary_device_capability()` with `_device_capability(gpu_id)`, cached per device via `@lru_cache(maxsize=8)`.
- Added bounds checks: non-int, negative, and out-of-range `gpu_id` values return `None` instead of raising.
- `_is_sm80(gpu_id)` now accepts a `gpu_id`.
- `_sm80_moe_runner_triton_override(gpu_id)` now accepts a `gpu_id`.
- `get_qwen3_omni_compat_overrides(gpu_id=None)` now accepts a `gpu_id`; `None` returns an empty dict for CPU stages.

Unchanged behavior:
- The only workaround is still `moe_runner_backend="triton"` on SM80.
- The escape hatch `SGLANG_OMNI_FORCE_SM80_MOE_TRITON=1` remains available.

### File 2: `sglang_omni/models/qwen3_omni/stages.py` (+41 / -1)

Added `_apply_compat_overrides(stage_name, gpu_id, overrides)`.

| Caller value | Stage GPU is SM80 | Behavior |
|---|---|---|
| unset | any | inject triton if a compat override exists |
| `"auto"` | any | inject triton |
| `"triton"` | any | preserve caller value |
| `"flashinfer_cutlass"` | yes | raise `ValueError` with a clear conflict message |
| `"flashinfer_cutlass"` | no | preserve caller value |
| other concrete backend | any | preserve caller value |

The `ValueError` path names the stage, `gpu_id`, configured backend, root cause, and two fixes: remove the explicit override or run on a different GPU.

Call sites:
- `create_sglang_thinker_executor_from_config`
- `create_talker_ar_executor_from_config`

Both call the helper after caller `server_args_overrides` are merged and before `ServerArgs` construction.

Unchanged:
- `config.py` is unchanged; Qwen3-Omni topology declarations stay as they were.
- Other models are unaffected.
- Qwen3-Omni text-only, colocated, and FP8 topologies are intended to remain unaffected.

---

## 3. Cross-Hardware Impact Matrix

| Hardware | Before | After |
|---|---|---|
| A100 (SM80) | works by forcing triton | works by auto-selecting triton |
| H100 / H20 (SM90) | could run triton unnecessarily | uses cutlass, matching upstream |
| B100 / B200 (SM100) | could run triton unnecessarily | uses cutlass |
| RTX 5090 / Blackwell consumer (SM120) | could run triton unnecessarily | uses cutlass |

After the fix, non-A100 hardware should be a compat no-op and continue through the upstream default `flashinfer_cutlass` or any higher-priority configured backend.

---

## 4. Contract

| Scenario | Behavior |
|---|---|
| A100 host, caller unset | inject `moe_runner_backend="triton"` |
| A100 host, caller set to `auto` | inject triton |
| A100 host, caller set to `triton` | preserve triton |
| A100 host, caller set to `flashinfer_cutlass` | raise `ValueError` explaining SM80 incompatibility and fixes |
| A100 host, caller set to `flashinfer_trtllm` | preserve caller value; SGLang owns any later validation |
| H100 host, any caller value | compat no-op; upstream/default behavior remains |
| mixed `cuda:0=H100`, `cuda:1=A100` | thinker on `cuda:0` no-op; talker on `cuda:1` injects triton |
| `gpu_id=None` | always no-op |
| invalid or out-of-range `gpu_id` | no-op, fail-safe |
| `SGLANG_OMNI_FORCE_SM80_MOE_TRITON=1` | force triton injection for bisecting or reproducing behavior |

---

## 5. Why Unit Tests Were Not Added

The P1/P2 fixes are contract-level changes, and the contract is documented directly in the `_apply_compat_overrides` docstring and the `compat.py` module docstring. Adding unit tests would increase the diff by roughly 180 lines. The current patch keeps the diff minimal and relies on the documented contract plus manual validation.

---

## 6. Validation

- Real A100 (8.0): overrides=`{"moe_runner_backend": "triton"}`
- Simulated H100 (9.0): overrides=`{}`
- Simulated B100 (10.0): overrides=`{}`
- Simulated Blackwell consumer (12.0): overrides=`{}`
- Simulated mixed `cuda:0=H100`, `cuda:1=A100`: `cuda:0={}`, `cuda:1=triton`
- Explicit `flashinfer_cutlass` on A100: `ValueError` fail-fast
- Explicit `flashinfer_cutlass` on H100: preserved
- Explicit `flashinfer_trtllm` on H100: preserved
- `gpu_id=None` and out-of-range `gpu_id=99`: no-op
- `SGLANG_OMNI_FORCE_SM80_MOE_TRITON=1` on H100: injects triton
- Real A100 end-to-end: 8-stage pipeline starts successfully, 1.3s end-to-end response
- `tests/unit_test/qwen3_omni/`: 257 tests pass with no regression

---

## 7. Difference From The Initial PR #711 Commit

| Item | Initial commit `a8826a1` | Current commit `5ab798f` |
|---|---|---|
| Total diff | +116 / -1 across 2 files | +123 / -1 across 2 files |
| Config changes | changed two `config.py` factory args | `config.py` is unchanged |
| Compat module | single `_is_sm80()` check | `_is_sm80(gpu_id)` with per-device cache and bounds checks |
| Override merge | `overrides.update(compat)` silently overwrote caller values | `_apply_compat_overrides()` with fail-fast contract |
| Explicit caller settings | silently overwritten | preserved, or rejected when incompatible |
| Comments | more verbose | minimal: docstring contract plus one inline call |
| Unit tests | none | none; contract is documented in code |

File: sglang_omni/models/qwen3_omni/stages.py [hunk 3]
```diff
@@ -59,6 +60,43 @@ class _ArMemoryContract:
...
+    gpu_id: int,
+    overrides: dict[str, Any],
+) -> None:
+    """Apply Qwen3-Omni hardware workarounds to ``overrides`` in place.
+
+    Rules: no-op on non-SM-80 GPUs; inject compat when caller is unset
+    or ``"auto"``; raise on caller-selected SM-80-incompatible backends;
+    otherwise preserve the caller's choice.
+    """
+    compat = _qwen3_omni_compat.get_qwen3_omni_compat_overrides(gpu_id)
```
Comment (Hayden727) [inline sglang_omni/models/qwen3_omni/stages.py:81]: [P1] Please add unit coverage for this compatibility contract instead of relying on the docstring plus manual validation. This helper now owns the safety-critical branches from the previous review: per-device SM80 detection, no-op on non-SM80 GPUs, injection only for unset/`auto`, preserving explicit `triton`, and fail-fast for explicit `flashinfer_cutlass` on SM80. A small test can monkeypatch `get_qwen3_omni_compat_overrides` / `_is_sm80` for `_apply_compat_overrides`, plus a focused fake-`torch` test for `compat.get_qwen3_omni_compat_overrides(gpu_id=0/1)`. Without that, the exact regression this PR fixes can come back with no CI signal.

**Review by Hayden727 — CHANGES_REQUESTED** [review]
The original two compatibility issues look addressed in this revision: the check is now keyed by `gpu_id`, and explicit SM80-incompatible backends fail fast instead of being silently overwritten. I still think this needs a small regression test before merge because the new helper is the entire contract boundary for A100/non-A100 placement and explicit backend precedence.

File: sglang_omni/models/qwen3_omni/stages.py (hunk 3 above)
Comment (YuTaoV5) [reply sglang_omni/models/qwen3_omni/stages.py:81]: Thanks — added tests/unit_test/qwen3_omni/test_compat.py with 6                                                                                                                                                                                            
     focused regression tests covering the three contract boundaries you                                                                                                                                                                                        
     called out:                                                                                                                                                                                                                                                
                                                                                                                                                                                                                                                                
     - per-gpu_id dispatch (test_mixed_host_dispatches_per_gpu_id)                                                                                                                                                                                              
     - explicit SM-80-incompatible backend precedence                                                                                                                                                                                                           
       (test_explicit_sm80_incompatible_backend_raises)                                                                                                                                                                                                         
     - explicit non-conflicting backends preserved                                                                                                                                                                                                              
       (test_explicit_other_backend_preserved_on_sm80, parametrized                                                                                                                                                                                             
       over triton / flashinfer_trtllm / cutlass)                                                                                                                                                                                                               
                                                                                                                                                                                                                                                                
     Full qwen3_omni test suite: 263 passed, no regressions.

File: tests/unit_test/qwen3_omni/test_compat.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+def _patch_torch_caps(*caps):
+    iterator = iter(caps)
+    fake = m.MagicMock()
+    fake.cuda.is_available.return_value = True
+    fake.cuda.device_count.return_value = len(caps)
+    fake.cuda.get_device_capability.side_effect = lambda i: next(iterator)
+    return m.patch.dict("sys.modules", {"torch": fake})
+
+
+def test_sm80_host_injects_triton():
```
Comment (Hayden727) [inline tests/unit_test/qwen3_omni/test_compat.py:31]: These SM80 assertions depend on the test runner's real CUDA capability. On H100/CPU/non-A100 hosts, `get_qwen3_omni_compat_overrides(gpu_id=0)` returns `{}`, and the later `_apply_compat_overrides` cases also no-op instead of injecting/raising, so this file only passes when CI happens to run on A100. Please wrap the SM80 cases with the fake torch capability patch (or monkeypatch the compat helpers for `_apply_compat_overrides`) so the regression is hardware-independent.

**Review by Hayden727 — CHANGES_REQUESTED** [review]
I rechecked the current head. The earlier placement and explicit-override issues look addressed, but the new regression tests still need one fix before merge.

**Review by Hayden727 — APPROVED** [review]
I rechecked the rebased head be9f8843. The SM80 compatibility contract is now hardware-independent in tests, the Qwen3-Omni AR factories still apply compat after caller overrides and before ServerArgs construction, and the branch is rebased onto current main c00e6f9d.\n\nValidated on hyper00-omni / sglang-omni-hayden-main-profile (H200, CUDA capability 9.0):\n- python -m py_compile sglang_omni/models/qwen3_omni/compat.py sglang_omni/models/qwen3_omni/stages.py tests/unit_test/qwen3_omni/test_compat.py\n- PYTHONPATH=/tmp/sglang_pr711_validate pytest -q tests/unit_test/qwen3_omni/test_compat.py -> 6 passed\n- CUDA_VISIBLE_DEVICES=0 PYTHONPATH=/tmp/sglang_pr711_validate pytest -q tests/unit_test/qwen3_omni -> 285 passed\n- Real H200 compat check returns {} as expected for non-SM80.
